mirrord/agent/src/dns.rs

```diff
@@ -75,7 +76,7 @@ impl DnsWorker {
     ///
     /// We could probably cache results here.
     /// We cannot cache the [`AsyncResolver`] itself, becaues the configuration in `etc` may change.
-    #[tracing::instrument(level = "trace")]
+    #[tracing::instrument(level = Level::TRACE, ret, err)]
```

Comment:
This should not be a global agent error imho. User puts bad url in their app -> error spam in the agent. I'd log error when we fail to build the resolver though. You can put building the resolver in a `try` block (excluding only the lookup itself) and log the error returned from that block

Reply:
Alright, for some reason the `try` block approach did not work. Rust could not figure out the type, and I tried putting the type in, both on the `let` side and on the `Ok` side, still kept complaining.

Using an `async` closure to the same effect.

Reply:
Tadaa. This compiles

```rust
    #[tracing::instrument(level = Level::TRACE, ret, err(level = Level::TRACE))]
    async fn do_lookup(
        etc_path: PathBuf,
        host: String,
        attempts: usize,
        timeout: Duration,
    ) -> RemoteResult<DnsLookup> {
        let resolver: Result<_, ResponseError> = try {
            let resolv_conf_path = etc_path.join("resolv.conf");
            let hosts_path = etc_path.join("hosts");

            let resolv_conf = fs::read(resolv_conf_path).await?;
            let hosts_conf = fs::read(hosts_path).await?;

            let (config, mut options) = parse_resolv_conf(resolv_conf)?;
            options.server_ordering_strategy =
                hickory_resolver::config::ServerOrderingStrategy::UserProvidedOrder;
            options.timeout = timeout;
            options.attempts = attempts;
            options.ip_strategy = hickory_resolver::config::LookupIpStrategy::Ipv4Only;

            let mut resolver = AsyncResolver::tokio(config, options);

            let hosts = Hosts::default().read_hosts_conf(hosts_conf.as_slice())?;
            resolver.set_hosts(Some(hosts));

            resolver
        };

        let lookup = resolver
            .inspect_err(|error| tracing::error!(%error, "failed to build DNS resolver"))?
            .lookup_ip(host)
            .await
            .inspect(|lookup| tracing::trace!(?lookup, "Lookup finished"))?
            .into();

        Ok(lookup)
    }
```

Reply:
Thank you!

Changed.